File: mobile/ios/MultimodalTrading/Views/Components/OrderBookView.swift
import SwiftUI

struct OrderBookView: View {
    @State private var bids: [OrderBookEntry] = []
    @State private var asks: [OrderBookEntry] = []
    @State private var isLoading: Bool = true
    let symbol: String
    
    var body: some View {
        VStack(spacing: 12) {
            Text("Order Book")
                .font(.headline)
                .frame(maxWidth: .infinity, alignment: .leading)
            
            if isLoading {
                ProgressView()
                    .frame(maxWidth: .infinity, minHeight: 100)
            } else {
                HStack(alignment: .top, spacing: 16) {
                    // Bids (Buy Orders)
                    VStack(spacing: 8) {
                        HStack {
                            Text("Price")
                                .font(.caption)
                                .foregroundColor(.gray)
                            
                            Spacer()
                            
                            Text("Amount")
                                .font(.caption)
                                .foregroundColor(.gray)
                        }
                        
                        ForEach(bids) { bid in
                            HStack {
                                Text(bid.price)
                                    .font(.caption)
                                    .foregroundColor(.green)
                                
                                Spacer()
                                
                                Text(bid.amount)
                                    .font(.caption)
                            }
                        }
                    }
                    .frame(maxWidth: .infinity)
                    
                    // Asks (Sell Orders)
                    VStack(spacing: 8) {
                        HStack {
                            Text("Price")
                                .font(.caption)
                                .foregroundColor(.gray)
                            
                            Spacer()
                            
                            Text("Amount")
                                .font(.caption)
                                .foregroundColor(.gray)
                        }
                        
                        ForEach(asks) { ask in
                            HStack {
                                Text(ask.price)
                                    .font(.caption)
                                    .foregroundColor(.red)
                                
                                Spacer()
                                
                                Text(ask.amount)
                                    .font(.caption)
                            }
                        }
                    }
                    .frame(maxWidth: .infinity)
                }
            }
        }
        .padding()
        .onAppear {
            loadOrderBook()
        }
        .onChange(of: symbol) { _ in
            loadOrderBook()
        }
    }
    
    // Load Order Book
    private func loadOrderBook() {
        isLoading = true
        
        // In a real implementation, this would call the API
        // For demo purposes, we'll generate mock data
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            // Generate mock order book data
            let basePrice = 103.42
            var mockBids: [OrderBookEntry] = []
            var mockAsks: [OrderBookEntry] = []
            
            // Generate bids (buy orders) slightly below base price
            for i in 0..<5 {
                let price = (basePrice - (Double(i) * 0.02) - (Double.random(in: 0...0.01))).formatted(.number.precision(.fractionLength(2)))
                let amount = (Double.random(in: 0.5...10.0)).formatted(.number.precision(.fractionLength(2)))
                mockBids.append(OrderBookEntry(id: UUID(), price: price, amount: amount))
            }
            
            // Generate asks (sell orders) slightly above base price
            for i in 0..<5 {
                let price = (basePrice + (Double(i) * 0.02) + (Double.random(in: 0...0.01))).formatted(.number.precision(.fractionLength(2)))
                let amount = (Double.random(in: 0.5...10.0)).formatted(.number.precision(.fractionLength(2)))
                mockAsks.append(OrderBookEntry(id: UUID(), price: price, amount: amount))
            }
            
            // Sort bids in descending order (highest price first)
            mockBids.sort { Double($0.price)! > Double($1.price)! }
            
            // Sort asks in ascending order (lowest price first)
            mockAsks.sort { Double($0.price)! < Double($1.price)! }
            
            self.bids = mockBids
            self.asks = mockAsks
            self.isLoading = false
        }
    }
}

// Order Book Entry Model
struct OrderBookEntry: Identifiable {
    let id: UUID
    let price: String
    let amount: String
}

struct OrderBookView_Previews: PreviewProvider {
    static var previews: some View {
        OrderBookView(symbol: "SOL/USDC")
            .previewLayout(.sizeThatFits)
            .padding()
    }
}

File: mobile/android/app/src/main/java/com/multimodaltrading/android/ui/components/OrderBookView.kt
package com.multimodaltrading.android.ui.components

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.multimodaltrading.android.viewmodels.TradingViewModel

@Composable
fun OrderBookView(
    symbol: String,
    modifier: Modifier = Modifier
) {
    var bids by remember { mutableStateOf<List<OrderBookEntry>>(emptyList()) }
    var asks by remember { mutableStateOf<List<OrderBookEntry>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }
    
    // Load order book data when symbol changes
    LaunchedEffect(symbol) {
        isLoading = true
        val orderBookData = loadOrderBook(symbol)
        bids = orderBookData.first
        asks = orderBookData.second
        isLoading = false
    }
    
    Card(
        modifier = modifier
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Text(
                text = "Order Book",
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(bottom = 8.dp)
            )
            
            if (isLoading) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(100.dp),
                    contentAlignment = androidx.compose.ui.Alignment.Center
                ) {
                    CircularProgressIndicator()
                }
            } else {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    // Bids (Buy Orders)
                    Column(
                        modifier = Modifier.weight(1f)
                    ) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween
                        ) {
                            Text(
                                text = "Price",
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                            
                            Text(
                                text = "Amount",
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        
                        Spacer(modifier = Modifier.height(8.dp))
                        
                        bids.forEach { bid ->
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween
                            ) {
                                Text(
                                    text = bid.price,
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.primary
                                )
                                
                                Text(
                                    text = bid.amount,
                                    style = MaterialTheme.typography.bodySmall
                                )
                            }
                            
                            Spacer(modifier = Modifier.height(4.dp))
                        }
                    }
                    
                    // Asks (Sell Orders)
                    Column(
                        modifier = Modifier.weight(1f)
                    ) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween
                        ) {
                            Text(
                                text = "Price",
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                            
                            Text(
                                text = "Amount",
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        
                        Spacer(modifier = Modifier.height(8.dp))
                        
                        asks.forEach { ask ->
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween
                            ) {
                                Text(
                                    text = ask.price,
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.error
                                )
                                
                                Text(
                                    text = ask.amount,
                                    style = MaterialTheme.typography.bodySmall
                                )
                            }
                            
                            Spacer(modifier = Modifier.height(4.dp))
                        }
                    }
                }
            }
        }
    }
}

// Load order book data
private fun loadOrderBook(symbol: String): Pair<List<OrderBookEntry>, List<OrderBookEntry>> {
    // In a real implementation, this would call the API
    // For demo purposes, we'll generate mock data
    
    val basePrice = when (symbol) {
        "SOL/USDC" -> 103.42
        "BTC/USDC" -> 62145.78
        "ETH/USDC" -> 3421.56
        else -> 100.0
    }
    
    val bids = List(5) { i ->
        val price = basePrice - (i * 0.02) - (Math.random() * 0.01)
        val amount = 0.5 + (Math.random() * 9.5)
        OrderBookEntry(
            price = String.format("%.2f", price),
            amount = String.format("%.2f", amount)
        )
    }.sortedByDescending { it.price.toDouble() }
    
    val asks = List(5) { i ->
        val price = basePrice + (i * 0.02) + (Math.random() * 0.01)
        val amount = 0.5 + (Math.random() * 9.5)
        OrderBookEntry(
            price = String.format("%.2f", price),
            amount = String.format("%.2f", amount)
        )
    }.sortedBy { it.price.toDouble() }
    
    return Pair(bids, asks)
}

// Order Book Entry data class
data class OrderBookEntry(
    val price: String,
    val amount: String
)
